library/src/blas1/reduction_strided_batched.h

```diff
@@ -0,0 +1,386 @@
+{
+    rocblas_int   tx = hipThreadIdx_x;
+    __shared__ To tmp[NB];
+
+    if(tx < nblocks)
+    {
+        To* work = workspace + hipBlockIdx_y * nblocks;
+        tmp[tx]  = work[tx];
+
+        // bound, loop
+        for(rocblas_int i = tx + NB; i < nblocks; i += NB)
+            REDUCE{}(tmp[tx], work[i]);
+    }
+    else
+    { // pad with default value
+        tmp[tx] = rocblas_default_value<To>{}();
+    }
+
+    if(nblocks < 32)
+    {
+        // no need parallel reduction
+        __syncthreads();
+
+        if(tx == 0)
+            for(rocblas_int i = 1; i < nblocks; i++)
+                REDUCE{}(tmp[0], tmp[i]);
+    }
+    else
+    {
+        // parallel reduction
+        rocblas_reduction<NB, REDUCE>(tx, tmp);
+    }
+
+    // Store result on device or in workspace
+    if(tx == 0)
+        result[hipBlockIdx_y] = FINALIZE{}(tmp[0]);
+}
+
+/*! \brief
+
+    \details
+    rocblas_reduction_strided_batched_kernel computes a reduction over multiple vectors x_i
+              Template parameters allow threads per block, data, and specific phase kernel overrides
+              At least two kernels are needed to finish the reduction
+              kernel 1 write partial result per thread block in workspace, blocks partial results
+              kernel 2 gathers all the partial result in workspace and finishes the final reduction.
+    @param[in]
+    handle    rocblas_handle.
+              handle to the rocblas library context queue.
+    @param[in]
+    n         rocblas_int
+              number of elements in each vector x_i
+    @param[in]
+    x         pointer to the first vector x_i on the GPU.
+    @param[in]
+    shiftx    rocblas_int
+              specifies a base offset increment for the start of each x_i.
+    @param[in]
+    incx      rocblas_int
+              specifies the increment for the elements of each x_i.
+    @param[in]
+    stridex   rocblas_int
+              specifies the pointer increment between batches for x. stridex must be >= n*incx
+    @param[in]
+    batch_count rocblas_int
+              number of instances in the batch
+    @param[out]
+    workspace To*  
+              temporary GPU buffer for inidividual block results for each batch
+              and results buffer in case result pointer is to host memory
+              Size must be (blocks+1)*batch_count*sizeof(To)
+    @param[out]
+    result
+              pointers to array of batch_count size for results. either on the host CPU or device GPU.
+              return is 0.0 if n, incx<=0.
+    ********************************************************************/
+template <rocblas_int NB,
+          typename Ti,
+          typename FETCH,
+          typename REDUCE   = rocblas_reduce_sum,
+          typename FINALIZE = rocblas_finalize_identity,
+          typename TPtrX,
+          typename To,
+          typename Tr>
+rocblas_status rocblas_reduction_strided_batched_kernel(rocblas_handle __restrict__ handle,
```

Comment:
I find it a bit confusing that these are named `_strided_batched` while they are used by all three scenarios. I would consider renaming - to me the suffix `_strided_batched` implies that it is tailored to `_strided_batched` functions, but maybe that's just me.